File: 652-find-duplicate-subtrees/find-duplicate-subtrees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_map<string,int> mp;
    vector<TreeNode*> ll;
    vector<TreeNode*> findDuplicateSubtrees(TreeNode* root) {
        dfs(root);
        return ll;
    }
    string dfs(TreeNode* root){
        if(!root){
            return "#";
        }
        string left=dfs(root->left);
        string right=dfs(root->right);
        string key=to_string(root->val)+","+left+","+right;
        mp[key]++;
        if(mp[key]==2){
            ll.push_back(root);
        }
        return key;
    }
};

File: 652-find-duplicate-subtrees/find-duplicate-subtrees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<TreeNode> findDuplicateSubtrees(TreeNode root) {
        HashMap<String,Integer> mp=new HashMap<>();
        List<TreeNode> ll=new ArrayList<>();
        serial(root,ll,mp);
        return ll;
    }
    public String serial(TreeNode root,List<TreeNode> ll,HashMap<String,Integer> mp){
        if(root==null){
            return "#";
        }
        String left=serial(root.left,ll,mp);
        String right=serial(root.right,ll,mp);
        String key=root.val+","+left+","+right;
        mp.put(key,mp.getOrDefault(key,0)+1);
        if(mp.get(key)==2){
            ll.add(root);
        }
        return key;
    }
}